refactoring: Rename LockAnnotation to LockAssertion and add run-time check to it

practicalswift opened the pull request:
`LockAnnotation lock(mutex);` is a guarantee to the compiler thread-analysis that `mutex` is locked (when it couldn't be determined otherwise).

Before this PR it was possible to make the mistake of adding a `LockAnnotation` where the correct mutex is _not_ held. This in turn makes the thread-analysis reasoning being based on incorrect premises.

This PR adds an assertion in the `LockAnnotation` ctor which checks that the guarantees given by us at compile-time are held also in practice (`ifdef DEBUG_LOCKORDER`).

Issues like the one described in #16028 will be discovered immediately with this PR merged.

Changes in this PR:
* Move `LockAnnotation` from `threadsafety.h` (imported code) to `sync.h` (our code)
* Move `LockAnnotation` in `wallet_tests` to make it reflect the truth
* Make sure the compile-time locking promises given via `LockAnnotation`:s hold also in practice at runtime (`ifdef DEBUG_LOCKORDER`)
* Rename `LockAnnotation` to `LockAssertion`

practicalswift committed:
Move LockAnnotation from threadsafety.h (imported code) to sync.h (our code)

practicalswift committed:
Make sure the compile-time locking promises given via LockAnnotation:s hold also in practice at runtime (ifdef DEBUG_LOCKORDER)

practicalswift committed:
Move LockAnnotation to make it reflect the truth

ryanofsky commented:
This change seems reasonable, but I don't think LockAnnotation is a good name for this class if it can potentially assert and abort the program. Would suggest adding a scripted diff to rename LockAnnotation to `LockAssert` or something similar.

practicalswift committed:
scripted-diff: Rename LockAnnotation to LockAssertion

-BEGIN VERIFY SCRIPT-
git grep -l LockAnnotation | xargs sed -i 's/LockAnnotation/LockAssertion/'
-END VERIFY SCRIPT-

practicalswift commented:
@ryanofsky Thanks for the quick review! I've now added a `scripted-diff` which renames `LockAnnotation` to `LockAssertion`. Please re-review :-)

promag commented:
Please update OP with the rename to `LockAssertion`.

practicalswift commented:
@promag Done! Please review :-)

ryanofsky reviewed (approved):
utACK last 4 commits of 8e81c1a87b45fea00bc0719569bd98ed61c1f6f4

practicalswift commented:
Rebased! Please re-review :-)

ryanofsky reviewed (approved):
utACK 9f85e9cb3d687862128ddf464d2bc2462b8627f0. No changes at all since last review except clean rebase after base PR #16033 was merged
